Integrate chronos

markstory opened the pull request:
This integrates cakephp/chronos in, and removes Carbon. In addition there are few other important changes that could be considered backwards incompatible, and at the very least will need to be documented:
- Date and DateTime columns return Immutable objects. This changes how one interacts with date/datetime objects. You can no longer `$date->modify('+1 minute')` for example. Instead you _must_ do `$date = $date->modify('+1 minute');`.
- A new `Cake\I18n\Date` class has been added. This class is used for Date columns, and presents the same interface as `Cake\I18n\Time`.
## Open Questions
- Are the changes related to Immutable objects something we are comfortable with?
- Should we include shims/aliases that allow userland code to continue to hint against `Carbon\Carbon`?
- Do the Date column changes seem like a risky change?

markstory committed:
Start integrating chronos.

* Update composer.json
* Remove methods defined in chronos.

markstory committed:
Start to integrate Chronos.

Extract a DateFormatTrait as soon there will be an Cake\I18n\Date class
to provide Time improvements for Date columns.

markstory committed:
Move additional code into the shared trait.

markstory committed:
Start adding the Date wrapper.

Add the I18n\Date class which adds the same i18n features to Dates that
we offer on Time instances. The timeAgoInWords() method has not been
implemented yet.

markstory committed:
Further integrate chronos into cakephp.

Finish integration with the database abstraction layer, and database
type mappers. The change to immutable datetimes may end up breaking some
applications, but hopefully not many.

markstory committed:
Accomodate immutable dates and use named formats.

Use the proper ISO-8601 formats.

markstory committed:
Add backwards compatibility shims

These methods are gross, but are required for backwards compatibility
with previous implementations of TimeHelper. This behavior is now
deprecated and will be removed in the next major release.

markstory committed:
Backup the locale property.

Don't leave global state in a worse place than when we found it.

markstory committed:
Implement incomplete tests for I18n\Date.

markstory committed:
Implement timeAgoInWords() for Date.

Duplicate many of the tests from Time, and remove tests related to time
components.

markstory committed:
Remove duplicate test.

antograssiot commented on src/I18n/Date.php:
```diff
@@ -0,0 +1,294 @@
+    protected static $_toStringFormat = [IntlDateFormatter::SHORT, -1];
+
+    /**
+     * The format to use when formatting a time using `Cake\I18n\Time::timeAgoInWords()`
```
using ``Cake\I18n\Date::timeAgoInWords()`` instead ?

lorenzo commented on src/Database/Type/DateTimeType.php:
```diff
@@ -126,7 +127,7 @@ public function toPHP($value, Driver $driver)
     public function marshal($value)
     {
-        if ($value instanceof DateTime) {
+        if ($value instanceof DateTime || $value instanceof DateTimeImmutable) {
```
DatetimeInterface will work better here

lorenzo commented:
My idea was to also offer a mutable `Time` implementation that people could opt into when upgrading. We can also document how to wire the `Type` classes so they retain the old behavior if necessary.

I think we need to add a shim for Carbon inside Chronos itself. That will also help attract people to the library.

Another thing I was thinking about is to make the `timeAgoInWords` implementation a separate object, that can be instantiated inside the date classes themselves. That way we can provide a migration path for user coming from Carbon that need their own implementation of `diffForHumans()`

lorenzo commented on tests/bootstrap.php:
```diff
@@ -101,6 +101,11 @@
+    //     'className' => 'Console',
+    //     'stream' => 'php://stderr',
+    //     'scopes' => ['queriesLog']
+    // ],
```
Was this intentional?

markstory replied:
Somewhat. I frequently enable query logging and wanted a quick way to enable it.

lorenzo replied:
Good stuff, then :)

markstory committed:
Incorporate feedback thus far.

markstory commented:
> I think we need to add a shim for Carbon inside Chronos itself. That will also help attract people to the library.

Ok, I can take care of that.

What did you have in mind for allowing mutable types? We could do something like:

```php
Type::build('datetime')->useImmutable();
Type::build('date')->useImmutable();
```

And default to mutable object creation. Defaulting to mutable objects would be more backwards compatible for existing applications. In the 3.2 compatible app skeleton we could default to immutable objects though.

markstory commented:
The test failures seem to be caused by intl versions, as the same tests pass fine for me locally. I'll try changing the test expectation around.

lorenzo commented:
@markstory I think the type classes have a setter (or maybe it was a public property?) for setting the datetime class to use. We can just point users to those options. I think the default skeleton does not  need to change as the type classes are now wired to the immutable classes directly

markstory commented:
While this pull request is defaulting to immutable dates, should we default to mutable ones for backwards compatibility reasons? Right now if an application is doing any mutations those changes will be lost. Any existing type checks will also fail. If we wanted to be more conservative, we could default to mutable, and enable immutability in the app-skeleton.

markstory committed:
See if german dates don't vary with intl versions.

lorenzo commented:
@markstory Being conservative would also meant that `Time` needs to extend from `MutableDateTime` instead of `Chronos`. But I think we can start having mutable dates in the Type classes for 3.2 and then change the default in 3.3?

markstory committed:
Default to mutable datetimes.

While I'd much prefer to use immutable date time objects. Historically,
we've provided userland code mutable instances. To keep compatibility
with existing use cases using mutable objects will be be safer.

markstory commented:
I've switched most of the code to be more compatible. `Cake\I18n\Time` is now inheriting from `MutableDateTime`. There are still a few failing tests related to dates not being mutable. I need to add a `MutableDate` to chronos for better compatibility, and will be doing that tomorrow.

markstory committed:
Use mutable date for Date base class.

Historically DateTimes from date columns were mutable.
Make that the default now as well.

jhli commented:
Why are mutable time objects bad? (Just asking)?

markstory committed:
Set testNow() on MutableDate too.

markstory commented:
@jhli The main benefit is that immutable objects are much harder to make mistakes with. For example:

```php
$thing->created->modify('+1 day');
```

Won't be persisted by the ORM right now, as the property didn't 'change'. Immutable objects make accidental side-effects impossible. The above example shows just one of the many ways datetimes could be modified in-place perhaps by accident. With an immutable object modifications are always explicit:

```php
$thing->created = $thing->created->modify('+1 day');
```

markstory committed:
Lets see if es-ES is more suitable for use in travis.ci

jhli commented:
Thanks - good point.

lorenzo commented:
Great job @markstory !

markstory commented:
Thanks @lorenzo :smile:

markstory commented:
I can look into providing immutable objects next.
